### PR TITLE
statistics: introduce cache interface for statsCache

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/34052

Problem Summary:

### What is changed and how it works?

introduce cache interface and useful struct for statsCache in order to use lru cache later.

### Check List

Tests 

- [x] Unit test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: statistics/histogram.go [hunk 1]
```diff
@@ -1091,15 +1091,31 @@ func (c *Column) GetIncreaseFactor(realtimeRowCount int64) float64 {
...
-func (c *Column) MemoryUsage() (sum int64) {
-	sum = c.Histogram.MemoryUsage()
+func (c *Column) MemoryUsage(tableMemUsage *TableMemoryUsage) (sum int64) {
+	var columnMemUsage *ColumnMemUsage
+	var ok bool
+	columnMemUsage, ok = tableMemUsage.ColumnsMemUsage[c.PhysicalID]
+	if !ok {
+		columnMemUsage = &ColumnMemUsage{
+			TableID:  tableMemUsage.TableID,
+			ColumnID: c.PhysicalID,
```
Comment (xuyifangreeneyes) [inline statistics/histogram.go:1103]: `c.PhysicalID` is the physical ID of the table. `c.Info.ID` is the column ID.

File: statistics/table.go [hunk 2]
```diff
@@ -112,21 +112,51 @@ type HistColl struct {
...
+	IndicesMemUsage map[int64]*IndexMemUsage
+}
+
+// ColumnMemUsage records column memory usage
+type ColumnMemUsage struct {
+	TableID           int64
+	ColumnID          int64
+	HistogramMemUsage int64
+	CMSketchMemUsage  int64
+	FMSketchMemUsage  int64
```
Comment (xuyifangreeneyes) [inline statistics/table.go:129]: The memory usage tracking of `TopN` in `Column` and that of `TopN` and `FMSketch` in `Index` seem missing. We can add it in the PR or another new PR.

Comment (Yisaer) [reply statistics/table.go:128]: The memUsage counting of `TopN` and `FMSketch` are not implemented now, I tended to implement them later.

File: statistics/handle/statscache.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+// Cost implements kvCache
+func (m *mapCache) Cost() int64 {
+	return m.memUsage
+}
+
+// Keys implements kvCache
+func (m *mapCache) Keys() []int64 {
+	ks := make([]int64, 0)
```
Comment (xuyifangreeneyes) [inline statistics/handle/statscache.go:146]: Can we preallocate memory for the slice?

File: statistics/handle/statscache.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	delete(m.tables, k)
+	m.memUsage -= item.cost
+}
+
+// Cost implements kvCache
+func (m *mapCache) Cost() int64 {
+	return m.memUsage
+}
+
+// Keys implements kvCache
```
Comment (xuyifangreeneyes) [inline statistics/handle/statscache.go:144]:
```suggestion
// Keys implements statsCacheInner
```

**Review by xuyifangreeneyes — COMMENTED** [review]
rest LGTM

File: statistics/handle/statscache.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+	// modified indeed. This can happen when we load extra column histograms into cache, or when we modify the cache with
+	// statistics feedbacks, etc. We cannot bump the version then because no new changes of `mysql.stats_meta` are loaded,
+	// while the override of statsCache is in a copy-on-write way, to make sure the statsCache is unchanged by others during the
+	// the interval of 'copy' and 'write', every 'write' should bump / check this minorVersion if the version keeps
+	// unchanged.
+	// This bump / check logic is encapsulated in `statsCache.update` and `updateStatsCache`, callers don't need to care
+	// about this minorVersion actually.
+	minorVersion uint64
+
+	statsCacheInner
```
Comment (chrysan) [inline statistics/handle/statscache.go:65]: Are you going to maintain LRU info for column stats in this structure?

Comment (Yisaer) [reply statistics/handle/statscache.go:65]: right

**Review by xuyifangreeneyes — APPROVED** [review]

**Review by JmPotato — APPROVED** [review]

**Review by chrysan — APPROVED** [review]
LGTM

**Pull request merged**